### DISCUSSION TITLE
How to detect that my function is invoked for completion

### DISCUSSION THREAD

**Discussion by delicb** [fish-shell/fish-shell#9192] [Q&A]
I have created a tool that automatically logs me in to various service (e.g. sets index url for `pip` or log me in to AWS via SSO). I wrap some of the functions to automatically perform login, if current credentials are expired, so that I can just use the tool and not worry about credentials. Example of function for `pip`:

```
function pip --wraps=pip --description "Wrapper for pip that does automatic sign in if needed"
	# if "org" command exists, use it to login in to pip
	# org command is smart enough not to login if pip credentials are
	# still valid, so it is not that expensive to call it every time
	if command -q org
		command org aws login --tool pip --quiet
	end
       command pip $argv
end
```
I have similar wrapper functions for `pipx`, `kubectl`, some are just invoking `org` with different arguments. 

This works fine. However, when interactively typing commands and fish attempts to generate autocomplete suggestion (at least that is my assumption), `org` command is executed. 

How can I detect that my function is executed in this context and skip calling `org` command? For autocomplete, it is not important to have fresh session and it can be annoying to have AWS SSO login flow started at random like this when I am in the middle of typing a command. 

I have tried to detect if there is `_PIP_COMPLETE` environment variable set, but no luck. Any suggestions?

---
**Update**

Huh, seems like every tool has environment variable for itself. 

`pip` is using `PIP_AUTO_COMPLETE`, `pipx` seems to be using https://github.com/kislyuk/argcomplete, so appropriate variable is `_ARGCOMPLETE`. I am using https://github.com/evanlucas/fish-kubectl-completions for `kubectl` and it seems to be working without any modifications. 

For `pip` (and equivalent for `pipx`) I have added the following at the start of the function:
```
	if set -q PIP_AUTO_COMPLETE
		command pip $argv
		return
	end
```

This works for me now. 

Is there a better way?

**Comment by faho** [comment]
>Huh, seems like every tool has environment variable for itself.

This is only the case for things that use a [click](https://github.com/pallets/click)-like autocompletion system. They expect to be started with a specific environment variable (which tbh always struck me as an awkward interface - I'd do it as `pip complete` instead, but oh well).

There isn't a general way to tell that you're inside a completion.

I'm going to recommend the same thing here I always recommend: When you make a function that does something fundamentally different from a specific command (other than, say, adding "--color=auto"), name it something else.

**Reply by delicb** [reply]
Thanks. 

In general, I do give functions names different from command itself, but in this case it seems to me that name should be the same, because my function "wraps" command and always executes it at the end, just ensures that environment is set up. Additionally, I have scripts that use `pip` and `kubectl` and I want those scripts to have benefit from the same behavior.

Thanks for confirmation that I am not missing anything regarding detection if invocation is completion related.
